### PR TITLE
Create test suite for bin/gem2rpm.

### PR DESCRIPTION
**Pull request by junaruga**
I created test suite for bin/gem2rpm.
This test suite must make us easier to modify the files, fix the bug, and add new features.
I think ideally that we can move the main logic from bin/gem2rpm to lib/.

# How to test

We can do usually like this.
```
rake test
```

If we wan to run the test suite in "non-internet environment" like Koji build in the Fedora project, 
We can set TEST_GEM2RPM_LOCAL environment variable.
The reason why I added TEST_GEM2RPM_LOCAL is that though I wanted to use mock feature for the logic that needs internet, it is hard to implement it, because its test will run on another process.

```
$ TEST_GEM2RPM_LOCAL=1 rake test
```

or

```
$ TEST_GEM2RPM_LOCAL=1 \
  ruby -Itest -e 'Dir.glob "./test/**/test_*.rb", &method(:require)'
```

# tmp directory
I am using tmp/[test_case_name]/ directory as a temporary directory for the each test case.
Why I used tmp/<test_case_name>/  is to run the case in parallel from minitest.

### CODE REVIEW

**Comment by junaruga** [comment]
I thought which way I will use for either mktemp for project_root/tmp/. And preferred project_root/ because I was afraid to create many directories under the /tmp, even if it will be removed on teardown method.
But I can change the logic to mktmp if you prefer mktemp.

File: .gitignore [hunk 1]
```diff
@@ -1 +1,2 @@
 .*
+/tmp
```
Comment (voxik) [inline .gitignore]: This is probably not what you want to ignore ...

File: test/helper.rb [hunk 2]
```diff
@@ -7,12 +8,30 @@
 
 require 'gem2rpm'
 
+# If you want to test in off line environment, set environment variable.
+def local_test?
```
Comment (voxik) [inline test/helper.rb]: Isn't there chance to rewrite this into method something like "skip_offline", so you could later avoid the conditions?

**Comment by junaruga** [comment]
Oh Travis CI test was failed. We need to install "rpmbuild" in the test environment.
https://travis-ci.org/fedora-ruby/gem2rpm/jobs/128864166

File: .gitignore (hunk 1 above)
Comment (junaruga) [inline .gitignore]: What is the correct way? 
My recognition is that "/tmp" matches for only PROJECT_ROOT/tmp directory, and "tmp" matches any directory under the PROJECT_ROOT such as PROJECT_ROOT/test/tmp.

File: test/helper.rb (hunk 2 above)
Comment (junaruga) [inline test/helper.rb]: okay I will create this kind of method.

```
def skip_if_offline
  skip(LOCAL_TEST_SKIPPING_MESSAGE) if local_test?
end
```

Comment (junaruga) [inline test/helper.rb]: Updated.

**Comment by junaruga** [comment]
On Travis (Ubuntu 12.04), the test was failed.

The reason is, because the gem2rpm is reading template file: default.spec.erb as default template on Travis.
And the generated RPM spec file has below features, and these causes the errors.

- Requires: ruby [""]
- URL: 

I am using Fedora23, and the gem2rpm used template: fedora-21-rawhide.spec.erb as a default template.
And generated RPM spec file has below features.
- "Requires: ruby [""]" is not created.
- If the Gem spec file's spec.homepage is empty, the RPM spec file's URL is not created.

See for detail: https://travis-ci.org/fedora-ruby/gem2rpm/jobs/128966588

Should I create template file for the Ubuntu?
or should I modify default.spec.erb?

**Comment by voxik** [comment]
You should use ```-t``` parameter to specify template. Or you can mock out the rpmbuild, which would be very likely better option anyway ...

File: .gitignore (hunk 1 above)
Comment (voxik) [inline .gitignore]: You are right of course. Sorry, my bad.

File: Gemfile [hunk 3]
```diff
@@ -2,5 +2,7 @@ source "https://www.rubygems.org"
 
 group :development do
   # Needed by Travis so we can run `bundle exec rake`
+  # Do not install rake, because it requires Ruby version >= 1.9
```
Comment (voxik) [inline Gemfile]: 1. This is not true
2. You probably don't want to submit such lines in your PR

Comment (junaruga) [inline Gemfile]: Thanks for the comment.
Firstly I had wanted to run 'bundle exec rake' on Travis.
And I added  ```gem 'rake'``` on Gemfile. Then I found it did not work on Travis.
So, I will remove this line and not submit in my PR.
But I think to write some comments for this line are useful to prevent someone to add same logic ```gem 'rake'```  like me in the future.

**Comment by junaruga** [comment]
I may repeat to push my branch to run my test suite on Travis from now.
Sorry for the disturbing you.

Let me tell you after fixing my test suite.

**Comment by junaruga** [comment]
Finally I could pass the tests on Travis, and on Red Hat Enterprise Linux 6.5 (Ruby 1.8.7, minitest 1.6.0) too.

Could you check this modification again?
Thanks!

High light
- .travis.yml
  - I needed to install rpmbuild for new test suite on Travis.
  - "cat /etc/os-release" is because it is affected to default template of gem2rpm.
    It looks better to display the information.
  - "rake TESTOPTS='-v'": We must want to know skpped test cases.
- bin/gem2rpm
  - Dir.exist? is not implemented on Ruby 1.8.7.
  - Added displaying the content of rpm spec file if rpmbuild was failed.
    It looks useful for debug.
- templates/ubuntu.spec.erb
  - It is for Travis. Travis is using ubuntu. Symbolic link to maintain easily.
- test/artifacts/testing_gem/testing_gem_error.gemspec
  - This is a gemspec file to test error cases using invalid gem file.
- test/helper.rb
  - Internal specification of "__ FILE __" is a little difference for between Ruby 1.8 and over than 1.9. I defined TEST_DIR to fix the test suite for all the targeting Ruby.

File: bin/gem2rpm [hunk 4]
```diff
@@ -81,6 +81,8 @@ if options[:srpm]
     "--define '_srcrpmdir #{out_dir}' #{specfile}"
   unless system(command)
     Gem2Rpm.show_message("Command failed: #{command}")
+    content = Gem2Rpm.file_content(specfile)
```
Comment (voxik) [inline bin/gem2rpm]: Sorry, but I don't like this behavior. There are plenty of other ways how to get the .spec file content.

Moreover, this is new feature and as such, it should not be part of this change.

File: lib/gem2rpm/package.rb [hunk 5]
```diff
@@ -1,4 +1,6 @@
 require 'delegate'
+# require 'rubygems' to avoid NameError at Ruby 1.8.7.
+require 'rubygems' unless defined?(Gem)
```
Comment (voxik) [inline lib/gem2rpm/package.rb]: What is the exact scenario this is needed?

File: templates/ubuntu.spec.erb [hunk 6]
```diff
@@ -0,0 +1 @@
+fedora-21-rawhide.spec.erb
```
Comment (voxik) [inline templates/ubuntu.spec.erb]: This is totally unacceptable. We don't want to ship any Ubuntu template.

File: test/helper.rb [hunk 7]
```diff
@@ -7,14 +9,46 @@
...
+  @is_local_test = ENV['TEST_GEM2RPM_LOCAL']
+end
+
+def skip_if_offline
+  skip('Skip fetch test in off line') if local_test?
```
Comment (voxik) [inline test/helper.rb]: This message is probably not universal enough.

File: lib/gem2rpm/package.rb (hunk 5 above)
Comment (junaruga) [inline lib/gem2rpm/package.rb]: It is below scenario.

https://travis-ci.org/fedora-ruby/gem2rpm/jobs/129158955
(GEM2RPM_DEBUG=1 is my private environment variable for my debug. Finally I deleted this variable with "git rebase".)

```
$ GEM2RPM_DEBUG=1 rake 
...
 10) Failure:
TestBinGem2Rpm#test_print_current_template [/home/travis/build/fedora-ruby/gem2rpm/test/test_bin_gem2rpm.rb:78]:
Command: /home/travis/build/fedora-ruby/gem2rpm/bin/gem2rpm -T
out: 
err: /home/travis/.rvm/rubies/ruby-1.8.7-p374/lib/ruby/site_ruby/1.8/rubygems/exceptions.rb:9: uninitialized constant Gem (NameError)
	from /home/travis/.rvm/rubies/ruby-1.8.7-p374/lib/ruby/site_ruby/1.8/rubygems/security.rb:7:in `require'
	from /home/travis/.rvm/rubies/ruby-1.8.7-p374/lib/ruby/site_ruby/1.8/rubygems/security.rb:7
	from /home/travis/.rvm/rubies/ruby-1.8.7-p374/lib/ruby/site_ruby/1.8/rubygems/package.rb:43:in `require'
	from /home/travis/.rvm/rubies/ruby-1.8.7-p374/lib/ruby/site_ruby/1.8/rubygems/package.rb:43
	from /home/travis/build/fedora-ruby/gem2rpm/lib/gem2rpm/package.rb:2:in `require'
	from /home/travis/build/fedora-ruby/gem2rpm/lib/gem2rpm/package.rb:2
	from /home/travis/build/fedora-ruby/gem2rpm/lib/gem2rpm.rb:3:in `require'
	from /home/travis/build/fedora-ruby/gem2rpm/lib/gem2rpm.rb:3
	from /home/travis/build/fedora-ruby/gem2rpm/bin/gem2rpm:6:in `require'
	from /home/travis/build/fedora-ruby/gem2rpm/bin/gem2rpm:6
```


The rpm spec file is below thing to prevent this error on RHEL 6.5, with ruby 1.8.7.
"require 'rubygems'" is called before test files are called.

```
ruby -Itest - << \EOF
  require 'rubygems'
  require 'minitest/unit'

  module Minitest; end;

  Minitest::Test = MiniTest::Unit::TestCase
  Dir.glob "./test/**/test_*.rb", &method(:require)
EOF
```

Comment (voxik) [inline lib/gem2rpm/package.rb]: Yes, and that is the point. The test fails, but I don't think you can trigger this behavior in real life usage, so I'd prefer to remove these two lines ...

Comment (junaruga) [inline lib/gem2rpm/package.rb]: Okay. I could check that real case passed without "require 'rubygems'" on RHEL 6.5, with ruby 1.8.7.

```
$ ruby -v
ruby 1.8.7 (2013-06-27 patchlevel 374) [x86_64-linux]

$ gem list

*** LOCAL GEMS ***

minitest (1.6.0)
rake (0.8.7)

$ bin/gem2rpm -T
...
- Initial package

$ echo $?
0

$ ruby lib/gem2rpm/package.rb

$ echo $?
0
```

However I am still confused.
Does its NameError happen only in the case of using minitest on ruby 1.8.7?
If it happens to that case, do we need to modify Rakefile or test files to pass the test (rake test) with ruby 1.8.7 on Travis? Because above error stack trace is for Travis ruby 1.8.7 case.

Comment (voxik) [inline lib/gem2rpm/package.rb]: No, it happens with more recent version of RubyGems then we are shipping by default in RHEL 6. I.e. there is RubyGems 2.5.1 used on Travis while we ship RubyGems 1.3.7 with RHEL 6.

But the thing is, that our users are typically using gem2rpm installed via RubyGems. In this case, the /usr/bin/gem2rpm executable generated by RubyGems ensures, that rubygems are loaded prior anything else no matter what is the RubyGems version. On more recent versions os Ruby are RubyGems loaded by default.

IOW by these two lines you are trying to solve scenario which is specific to Travis in this specific configuration and the case when one would run the gem2rpm directly from the checked out repository using obsolete version of Ruby. Nothing worth of the line.

Comment (junaruga) [inline lib/gem2rpm/package.rb]: I understood it. Thank you for your detail explanation.

**Comment by junaruga** [comment]
Hi, @voxik 
I will close this pull request, because I decided to send the several pull-requests with small content that are divided from this pull-request with content.
Obviously this pull request's conversation was useful for me.
Thank you for your taking the time, and reviewing for me.
The divided pull-requests with small content looks much better for developers, users, you and me.

So, I would to like to close this pull-request finally.
